tests/vmi_headless_test.go

```diff
@@ -163,6 +170,88 @@ var _ = Describe("[rfe_id:609][sig-compute]VMIheadless", func() {
 				Expect(vncCount).To(Equal(1), "should have exactly one VNC device")
 			})
 
+			It("multiple HTTP calls should re-use connections and not grow the number of open connections in virt-launcher", func() {
+				getHandlerConnectionCount := func() int {
+					cmd := []string{"bash", "-c", fmt.Sprintf("ss -ntlap | grep %d | wc -l", virt_api.DefaultConsoleServerPort)}
+					stdout, stderr, err := tests.ExecuteCommandOnNodeThroughVirtHandler(virtClient, vmi.Status.NodeName, cmd)
+					Expect(err).ToNot(HaveOccurred())
+					Expect(stderr).To(BeEmpty())
+
+					stdout = strings.TrimSpace(stdout)
+					stdout = strings.ReplaceAll(stdout, "\n", "")
+
+					handlerCons, err := strconv.Atoi(stdout)
+					Expect(err).ToNot(HaveOccurred())
+
+					return handlerCons
+				}
+
+				getClientCalls := func(vmi *v1.VirtualMachineInstance) []func() {
+					vmiInterface := virtClient.VirtualMachineInstance(vmi.ObjectMeta.Namespace)
+					expectNoErr := func(err error) {
+						ExpectWithOffset(2, err).ToNot(HaveOccurred())
+					}
+
+					return []func(){
+						func() {
+							_, err := vmiInterface.GuestOsInfo(vmi.Name)
+							expectNoErr(err)
+						},
+						func() {
+							_, err := vmiInterface.FilesystemList(vmi.Name)
+							expectNoErr(err)
+						},
+						func() {
+							_, err := vmiInterface.UserList(vmi.Name)
+							expectNoErr(err)
+						},
+						func() {
+							_, err := vmiInterface.VNC(vmi.Name)
+							expectNoErr(err)
+						},
+						func() {
+							_, err := vmiInterface.SerialConsole(vmi.Name, &kubecli.SerialConsoleOptions{ConnectionTimeout: 30 * time.Second})
+							expectNoErr(err)
+						},
+					}
+				}
+
+				By("Running the VMI")
+				vmi = tests.NewRandomFedoraVMIWithGuestAgent()
+				vmi = tests.RunVMIAndExpectLaunch(vmi, 30)
+
+				By("VMI has the guest agent connected condition")
+				Eventually(func() []v1.VirtualMachineInstanceCondition {
+					vmi, err = virtClient.VirtualMachineInstance(util.NamespaceTestDefault).Get(vmi.Name, &v12.GetOptions{})
+					Expect(err).ToNot(HaveOccurred(), "Should get VMI ")
+					return vmi.Status.Conditions
+				}, 30*time.Second, 2).Should(
+					ContainElement(
+						gstruct.MatchFields(
+							gstruct.IgnoreExtras,
+							gstruct.Fields{"Type": Equal(v1.VirtualMachineInstanceAgentConnected)})),
+					"Should have agent connected condition")
+
+				origHandlerCons := getHandlerConnectionCount()
+
+				By("Making multiple requests")
+				const numberOfRequests = 20
+				for i := 0; i < numberOfRequests; i++ {
+					clientCalls := getClientCalls(vmi)
+					for _, clientCallFunc := range clientCalls {
+						clientCallFunc()
+					}
+					time.Sleep(200 * time.Millisecond)
+				}
+
+				// Since this test does not run serially (but in parallel), it's possible that other tests would establish
+				// some connections to virt-handler. If this test would have run serially, this number should have been 1.
+				const extraConnectionsAllowed = 5
```

Comment:
Must have missed that during the previous review round... Maybe let's run the test serially then? I am not very confident that it will not produce flakes at some point.

Reply:
Yeah well I didn't want to hurt parallelism but on second thought, ensuring flakieness doesn't occur seems more important. Thanks for the advice. Done.